source/agora/node/Ledger.d

```diff
@@ -129,10 +129,25 @@ public class Ledger
         if (this.utxo_set.length == 0)
         {
             Block block;
+            Block last_block = gen_block;
             foreach (height; 0 .. this.last_block.header.height + 1)
             {
                 this.storage.readBlock(block, Height(height));
+
+                if (height != 0)
+                {
+                    size_t active_enrollments = enroll_man.getValidatorCount(
+                        block.header.height);
+
+                    if (auto fail_reason = block.isInvalidReason(last_block.header.height,
+                        last_block.header.hashFull,
+                        this.utxo_set.getUTXOFinder(),
+                        active_enrollments))
+                        throw new Exception("Rejected read block: " ~ fail_reason);
+                }
                 this.updateUTXOSet(block);
+                this.updateValidatorSet(block);
```

Comment:
On line 168 we are again rebuilding the validator set. I wonder if we can avoid doing it twice.

Additionally, @linked0 we seem to have both `restoreValidators` and `updateValidatorSet`. It seems that either one can be used, according to this code and line 172 below. Do you think we should simplify this and only have one function? Or is the code here wrong and it should use `restoreValidators`?

Reply:
Right. We already are rebuilding the validator set using the latest 1008 of blocks in `restoreValidators` on line 158.  So we don't need to call the  `updateValidatorSet` function here.  And, those two functions @AndrejMitrovic said really have different operations.

Reply:
No I think it's definitely needed here. What I'm saying is that *if* its done here, we can skip the block on line 158.

Reply:
updateValidatorSetThis may be necessary.

Because it violates block verification logic.
https://github.com/bpfkorea/agora/blob/dc869bd16b84188606f3c514b4aefb197e7c26e3/source/agora/consensus/validation/Block.d#L111-L112
I think we need to integrate with restoreValidatorslogic below.

Reply:
@AndrejMitrovic @TrustHenry You're right. We can remove the lines from 153 to 159. And we can remove `restoreValidators' also from `EnrollmentManager`. Sorry for the confusion!

Reply:
There are several cases we need to cover:
- utxo set is empty -> clear and rebuild both utxo set and validator set. Clearing is important because you might have an empty utxo set but a non-empty validator set
- only validator set is empty -> only rebuild validator set

So I don't think it's correct to remove that code.